rule_linux.go

```diff
@@ -297,7 +298,8 @@ func (h *Handle) RuleListFiltered(family int, filter *Rule, filterMask uint64) (
 				continue
 			case filterMask&RT_FILTER_MARK != 0 && rule.Mark != filter.Mark:
 				continue
-			case filterMask&RT_FILTER_MASK != 0 && rule.Mask != filter.Mask:
+			case filterMask&RT_FILTER_MASK != 0 &&
+				(rule.Mask != filter.Mask && (rule.Mask == nil || filter.Mask == nil || *rule.Mask != *filter.Mask)):
```

Comment:
I think this is correct, but I personally find it hard to read.
IMO, it's good to have a utility function for this. Something like this:
```go
func ptrEqual[T comparable](a, b *T) bool {
	if (a == nil) != (b == nil) {
		return false
	}
	if a == nil {
		return true
	}
	return *a == *b
}
```
Source: https://github.com/kubernetes/utils/blob/18e509b52bc8/ptr/ptr.go#L65

And then this case can become: `case filterMask&RT_FILTER_MASK != 0 && !ptrEqual(rule.Mask, filter.Mask)`

Reply:
Agree.
On a similar line, I prefer
```
func ptrEqual(int *a, int* b) bool {
        if a == b {
		return true
	}
	if (a == nil) || (b == nil) {
		return false
	}
	return *a == *b
}
which would avoid a deeper comparison if the caller passed the same pointer for both input parameters.
```

Reply:
I have add func `ptrEqual `